File: android-sdk/security-sdk/src/main/kotlin/com/bankingsdk/security/SecuritySdk.kt
package com.bankingsdk.security

import android.content.Context
import android.content.pm.ApplicationInfo
import android.os.Build
import android.os.Debug
import android.util.Base64
import androidx.annotation.Keep
import com.bankingsdk.security.crypto.AndroidKeystoreSigner
import com.bankingsdk.security.crypto.EcP256
import com.bankingsdk.security.crypto.HybridEnvelope
import java.io.File
import java.net.URI
import org.json.JSONObject

/**
 * Core Android security SDK facade. Cryptography runs only in native code; private keys stay in Keystore.
 */
@Keep
object SecuritySdk {

    @Volatile
    private var appContext: Context? = null

    @Volatile
    private var serverSpki: ByteArray? = null

    @Volatile
    private var keystoreSigner: AndroidKeystoreSigner? = null

    /**
     * Initializes hardware-backed signing (EC P-256, non-exportable) and encrypted installation identity.
     */
    fun init(context: Context): Result<Unit> = runCatching {
        val app = context.applicationContext
        appContext = app
        keystoreSigner = AndroidKeystoreSigner()
        DeviceIdentity.getOrCreateInstallationId(app)
    }

    fun configureServerPublicKey(spkiDer: ByteArray): Result<Unit> = runCatching {
        EcP256.publicKeyFromSpki(spkiDer)
        serverSpki = spkiDer.copyOf()
    }

    fun configureServerPublicKeyFromBase64(b64: String): Result<Unit> {
        val bytes = Base64.decode(b64, Base64.NO_WRAP)
        return configureServerPublicKey(bytes)
    }

    /** ECDSA P-256 signing public key (SPKI DER) as Base64 — for device registration. */
    fun getPublicKey(): Result<String> = runCatching {
        appContext ?: error("init(context) required")
        val signer = keystoreSigner ?: AndroidKeystoreSigner().also { keystoreSigner = it }
        Base64.encodeToString(signer.signingPublicKeySpki(), Base64.NO_WRAP)
    }

    /**
     * Rotates the hardware-backed signing key. Prior key material is invalidated in Keystore.
     */
    fun rotateKeys(): Result<Unit> = runCatching {
        appContext ?: error("init(context) required")
        val signer = keystoreSigner ?: AndroidKeystoreSigner().also { keystoreSigner = it }
        signer.rotateSigningKeys()
    }

    /**
     * Payload for server-side device registration (installation id + signing public key + platform).
     */
    fun getDeviceRegistrationPayload(): Result<Map<String, Any?>> = runCatching {
        val ctx = appContext ?: error("init(context) required")
        val signer = keystoreSigner ?: AndroidKeystoreSigner().also { keystoreSigner = it }
        val installationId = DeviceIdentity.getOrCreateInstallationId(ctx)
        val pubB64 = Base64.encodeToString(signer.signingPublicKeySpki(), Base64.NO_WRAP)
        mapOf(
            "installationId" to installationId,
            "signingPublicKeySpki" to pubB64,
            "platform" to "android",
        )
    }

    fun secureRequest(path: String, body: Map<String, Any?>): Result<SecureResponse> = runCatching {
        val ctx = appContext ?: error("init(context) required")
        val spki = serverSpki ?: error("configureServerPublicKey first")
        val signer = keystoreSigner ?: AndroidKeystoreSigner().also { keystoreSigner = it }
        val plain = bodyToJsonBytes(body)
        val (host, p) = parseHostAndPath(path)
        val risk = getSecurityStatus().riskScore
        val env = HybridEnvelope.build(
            method = "POST",
            host = host,
            contentType = "application/json",
            riskScore = risk,
            path = p,
            plaintext = plain,
            serverPublicSpki = spki,
            signer = signer,
        )
        SecureResponse(
            statusCode = 200,
            headers = mapOf("Content-Type" to "application/json"),
            body = env,
        )
    }

    fun sign(data: ByteArray): Result<ByteArray> = runCatching {
        val ctx = appContext ?: error("init(context) required")
        val signer = keystoreSigner ?: AndroidKeystoreSigner().also { keystoreSigner = it }
        signer.sign(data)
    }

    /** Stable installation identifier (encrypted at rest), for registration — not raw ANDROID_ID. */
    fun getDeviceId(): Result<String> = runCatching {
        val ctx = appContext ?: error("init(context) required")
        DeviceIdentity.getOrCreateInstallationId(ctx)
    }

    data class SecurityStatus(
        /** 0..100 (higher = riskier runtime). */
        val riskScore: Int,
        /** Array of stable string codes describing findings. */
        val findings: List<String>,
    )

    fun getSecurityStatus(): SecurityStatus {
        val ctx = appContext
        val findings = mutableListOf<String>()
        var score = 0

        val emulator = isProbablyEmulator()
        if (emulator) {
            findings.add("emulator")
            score += 30
        }

        val rooted = isProbablyRooted()
        if (rooted) {
            findings.add("root")
            score += 45
        }

        val debug = Debug.isDebuggerConnected() || Debug.waitingForDebugger()
        if (debug) {
            findings.add("debugger")
            score += 40
        }

        if (ctx != null && (ctx.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0) {
            findings.add("debuggable_app")
            score += 10
        }

        val hooks = hasHookingIndicators()
        if (hooks) {
            findings.add("hooks")
            score += 50
        }

        return SecurityStatus(
            riskScore = score.coerceIn(0, 100),
            findings = findings.distinct(),
        )
    }

    fun keyManager(): SecureKeyManager? {
        val ctx = appContext ?: return null
        return SecureKeyManager(ctx)
    }

    data class SecureResponse(
        val statusCode: Int,
        val headers: Map<String, String>,
        val body: Map<String, Any?>,
    )

    private fun bodyToJsonBytes(body: Map<String, Any?>): ByteArray {
        val o = JSONObject()
        body.forEach { (k, v) ->
            when (v) {
                null -> o.put(k, JSONObject.NULL)
                is Map<*, *> -> throw IllegalArgumentException("nested maps not supported")
                is List<*> -> o.put(k, org.json.JSONArray(v))
                else -> o.put(k, v)
            }
        }
        return o.toString().toByteArray(Charsets.UTF_8)
    }

    private fun parseHostAndPath(input: String): Pair<String, String> {
        return try {
            val u = URI(input)
            val host = u.host ?: ""
            val path = buildString {
                append(u.rawPath ?: "")
                if (!u.rawQuery.isNullOrBlank()) append("?").append(u.rawQuery)
            }.ifBlank { input }
            host to path
        } catch (_: Throwable) {
            "" to input
        }
    }

    private fun isProbablyEmulator(): Boolean {
        val fingerprint = Build.FINGERPRINT.lowercase()
        val model = Build.MODEL.lowercase()
        val manufacturer = Build.MANUFACTURER.lowercase()
        val brand = Build.BRAND.lowercase()
        val device = Build.DEVICE.lowercase()
        val product = Build.PRODUCT.lowercase()

        return fingerprint.startsWith("generic") ||
            fingerprint.contains("vbox") ||
            fingerprint.contains("test-keys") ||
            model.contains("google_sdk") ||
            model.contains("emulator") ||
            model.contains("android sdk built for") ||
            manufacturer.contains("genymotion") ||
            brand.startsWith("generic") && device.startsWith("generic") ||
            product.contains("sdk") ||
            product.contains("emulator") ||
            product.contains("simulator")
    }

    private fun isProbablyRooted(): Boolean {
        // Fast heuristics; no single check is definitive.
        val tags = Build.TAGS ?: ""
        if (tags.contains("test-keys")) return true

        val paths = arrayOf(
            "/system/app/Superuser.apk",
            "/sbin/su",
            "/system/bin/su",
            "/system/xbin/su",
            "/data/local/xbin/su",
            "/data/local/bin/su",
            "/system/sd/xbin/su",
            "/system/bin/failsafe/su",
            "/data/local/su",
        )
        if (paths.any { File(it).exists() }) return true

        return try {
            val p = Runtime.getRuntime().exec(arrayOf("su", "-c", "id"))
            val rc = p.waitFor()
            rc == 0
        } catch (_: Throwable) {
            false
        }
    }

    private fun hasHookingIndicators(): Boolean {
        // Look for known strings in loaded memory mappings (Frida/Xposed/Substrate).
        return try {
            val maps = File("/proc/self/maps")
            if (!maps.exists()) return false
            val txt = maps.readText()
            val t = txt.lowercase()
            t.contains("frida") ||
                t.contains("gadget") && t.contains("frida") ||
                t.contains("xposed") ||
                t.contains("substrate") ||
                t.contains("libhook") ||
                t.contains("magisk")
        } catch (_: Throwable) {
            false
        }
    }
}

File: ios-sdk/Sources/SecuritySDK/SecuritySDK.swift
import CryptoKit
import Darwin
import Foundation
import Security

/// Native-only security facade (Secure Enclave / Keystore-grade signing). No JavaScript.
public enum SecuritySdk {
    private static var serverSpki: Data?
    private static let signer = KeychainSigner()

    /// Initializes hardware-backed signing (Secure Enclave when available) and installation identity.
    public static func initSdk() throws {
        try signer.warmUp()
        _ = try InstallationIdentity.getOrCreateId()
    }

    public static func configureServerPublicKey(spkiDer: Data) throws {
        _ = try EcSpki.x963FromSecp256r1Spki(spkiDer)
        serverSpki = spkiDer
    }

    public static func configureServerPublicKeyFromBase64(_ b64: String) throws {
        guard let data = Data(base64Encoded: b64) else {
            throw NSError(domain: "SecuritySdk", code: 1, userInfo: [NSLocalizedDescriptionKey: "invalid base64"])
        }
        try configureServerPublicKey(spkiDer: data)
    }

    /// ECDSA P-256 signing public key (SPKI DER), Base64 — for device registration.
    public static func getPublicKey() throws -> String {
        try signer.signingPublicKeySpki().base64EncodedString()
    }

    /// Rotates the signing key (invalidates prior material in Secure Enclave / Keychain).
    public static func rotateKeys() throws {
        try signer.rotateKeys()
    }

    /// Values to send to your registration endpoint.
    public static func getDeviceRegistrationPayload() throws -> [String: Any] {
        let installationId = try InstallationIdentity.getOrCreateId()
        let pubB64 = try signer.signingPublicKeySpki().base64EncodedString()
        return [
            "installationId": installationId,
            "signingPublicKeySpki": pubB64,
            "platform": "ios",
        ]
    }

    public struct SecureResponse {
        public let statusCode: Int
        public let headers: [String: String]
        public let body: [String: Any]

        public init(statusCode: Int, headers: [String: String], body: [String: Any]) {
            self.statusCode = statusCode
            self.headers = headers
            self.body = body
        }
    }

    public static func secureRequest(path: String, body: [String: Any]) throws -> SecureResponse {
        guard let spki = serverSpki else {
            throw NSError(domain: "SecuritySdk", code: 2, userInfo: [NSLocalizedDescriptionKey: "configure server public key first"])
        }
        let plain = try JSONSerialization.data(withJSONObject: body, options: [])
        let (host, p) = parseHostAndPath(path)
        let risk = getSecurityStatus().riskScore
        let env = try HybridEnvelopeIOS.build(
            method: "POST",
            host: host,
            contentType: "application/json",
            riskScore: risk,
            path: p,
            plaintext: plain,
            serverPublicSpki: spki,
            signer: signer
        )
        return SecureResponse(statusCode: 200, headers: ["Content-Type": "application/json"], body: env)
    }

    public static func sign(data: Data) throws -> Data {
        try signer.sign(data: data)
    }

    /// Stable installation id (Keychain), for registration — not a raw vendor id.
    public static func getDeviceId() throws -> String {
        try InstallationIdentity.getOrCreateId()
    }

    public struct SecurityStatus {
        /// 0..100 (higher = riskier runtime).
        public let riskScore: Int
        /// Array of stable string codes describing findings.
        public let findings: [String]

        public init(riskScore: Int, findings: [String]) {
            self.riskScore = max(0, min(100, riskScore))
            self.findings = Array(Set(findings))
        }
    }

    public static func getSecurityStatus() -> SecurityStatus {
        var findings: [String] = []
        var score = 0

        let isSim = ProcessInfo.processInfo.environment["SIMULATOR_DEVICE_NAME"] != nil
        if isSim {
            findings.append("simulator")
            score += 30
        }

        if isDebuggerAttached() {
            findings.append("debugger")
            score += 40
        }

        if isProbablyJailbroken() {
            findings.append("jailbreak")
            score += 55
        }

        if hasHookingIndicators() {
            findings.append("hooks")
            score += 50
        }

        return SecurityStatus(riskScore: score, findings: findings)
    }

    public static func keyManagerPublicSpki() throws -> Data {
        try signer.signingPublicKeySpki()
    }
}

private func parseHostAndPath(_ input: String) -> (String, String) {
    guard let url = URL(string: input), let host = url.host else {
        return ("", input)
    }
    let path = (url.path.isEmpty ? "/" : url.path) + (url.query.map { "?\($0)" } ?? "")
    return (host, path)
}

private func isDebuggerAttached() -> Bool {
    var info = kinfo_proc()
    var mib: [Int32] = [CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()]
    var size = MemoryLayout<kinfo_proc>.stride
    let rc = sysctl(&mib, u_int(mib.count), &info, &size, nil, 0)
    guard rc == 0 else { return false }
    return (info.kp_proc.p_flag & P_TRACED) != 0
}

private func isProbablyJailbroken() -> Bool {
    #if targetEnvironment(simulator)
    return false
    #else
    let fm = FileManager.default
    let suspiciousPaths = [
        "/Applications/Cydia.app",
        "/Library/MobileSubstrate/MobileSubstrate.dylib",
        "/bin/bash",
        "/usr/sbin/sshd",
        "/etc/apt",
        "/private/var/lib/apt/",
    ]
    if suspiciousPaths.contains(where: { fm.fileExists(atPath: $0) }) {
        return true
    }

    // Attempt to write outside the sandbox (should fail on stock devices).
    let testPath = "/private/jb_\(UUID().uuidString)"
    do {
        try "x".write(toFile: testPath, atomically: true, encoding: .utf8)
        try fm.removeItem(atPath: testPath)
        return true
    } catch {
        // Expected.
    }

    return false
    #endif
}

private func hasHookingIndicators() -> Bool {
    // Scan loaded dynamic libraries for common injection names.
    let suspectSubstrings = [
        "frida",
        "fridagadget",
        "cydia",
        "substrate",
        "mobilesubstrate",
        "libhooker",
        "fishhook",
        "cycript",
        "xposed",
    ]

    let imageCount = _dyld_image_count()
    for i in 0..<imageCount {
        guard let cName = _dyld_get_image_name(i) else { continue }
        let name = String(cString: cName).lowercased()
        if suspectSubstrings.contains(where: { name.contains($0) }) {
            return true
        }
    }

    if let env = getenv("DYLD_INSERT_LIBRARIES") {
        let s = String(cString: env)
        if !s.isEmpty {
            return true
        }
    }

    return false
}
